pkg/etcdcli/etcdcli.go

```diff
@@ -297,39 +298,74 @@ func (g *etcdClientGetter) UnhealthyMembers() ([]*etcdserverpb.Member, error) {
 	defer cancel()
 
-	membersResp, err := cli.MemberList(ctx)
+	etcdCluster, err := cli.MemberList(ctx)
 	if err != nil {
 		return nil, err
 	}
 
-	unhealthyMembers := []*etcdserverpb.Member{}
-	unstartedMemberNames := []string{}
-	unhealthyMemberNames := []string{}
-	for _, member := range membersResp.Members {
-		if len(member.ClientURLs) == 0 {
-			unhealthyMembers = append(unhealthyMembers, member)
-			unstartedMemberNames = append(unstartedMemberNames, GetMemberNameOrHost(member))
-			continue
-		}
-		ctx, cancel := context.WithCancel(context.Background())
-		defer cancel()
-
-		_, err := cli.Status(ctx, member.ClientURLs[0])
-		if err != nil {
-			unhealthyMembers = append(unhealthyMembers, member)
-			unhealthyMemberNames = append(unhealthyMemberNames, member.Name)
-		}
-	}
-
+	etcdMemberHealth := g.getMemberHealth(etcdCluster.Members)
+	unstartedMemberNames := etcdMemberHealth.GetUnstartedMemberNames()
 	if len(unstartedMemberNames) > 0 {
 		g.eventRecorder.Warningf("UnstartedEtcdMember", "unstarted members: %v", strings.Join(unstartedMemberNames, ","))
 	}
 
+	unhealthyMemberNames := etcdMemberHealth.GetUnhealthyMemberNames()
 	if len(unhealthyMemberNames) > 0 {
 		g.eventRecorder.Warningf("UnhealthyEtcdMember", "unhealthy members: %v", strings.Join(unhealthyMemberNames, ","))
 	}
 
-	return unhealthyMembers, nil
+	return etcdMemberHealth.GetUnhealthyMembers(), nil
+}
+
+func (g *etcdClientGetter) MemberHealth(etcdMembers []*etcdserverpb.Member) *memberHealth {
+	return g.getMemberHealth(etcdMembers)
+}
+
+func (g *etcdClientGetter) getMemberHealth(etcdMembers []*etcdserverpb.Member) *memberHealth {
+	var wg sync.WaitGroup
+	memberHealth := &memberHealth{}
+	hch := make(chan healthCheck, len(etcdMembers))
+	for _, member := range etcdMembers {
+		if len(member.ClientURLs) == 0 {
+			memberHealth.Check = append(memberHealth.Check, healthCheck{Member: member, Health: false, Started: false})
+			continue
+		}
+		wg.Add(1)
+		go func(member *etcdserverpb.Member) {
+			defer wg.Done()
+			// new client vs shared is used here to minimize disruption of cached client consumers.
+			// performance analisis of CPU and RSS consumption showed net gain after refactor
+			cli, err := getEtcdClient([]string{member.ClientURLs[0]})
+			if err != nil {
+				hch <- healthCheck{Member: member, Health: false, Started: true, Error: fmt.Errorf("create client failure: %w", err)}
+				return
+			}
+			defer cli.Close()
+			st := time.Now()
+			ctx, cancel := context.WithCancel(context.Background())
+			// linearized request to verify health of member
+			_, err = cli.Get(ctx, "health")
+			cancel()
+			hc := healthCheck{Member: member, Health: false, Started: true, Took: time.Since(st).String()}
+			if err == nil || err == rpctypes.ErrPermissionDenied {
+				hc.Health = true
+			} else {
+				hc.Error = fmt.Errorf("health check failed: %w", err)
+			}
+			hch <- hc
+		}(member)
+	}
+
+	go func() {
+		wg.Wait()
```

Comment:
I wonder why Wait() is used in a goroutine.
Only after wg.Wait() returns, would we be able to read all healthCheck's from the channel.

Reply:
Agree, this seems like an error. Because the channel is buffered, I think the `range` can cause the function to return before all items are read because the channel read will not block if the channel is empty (so the `range` consumer races with the producer goroutine).

I think the `Wait()`/`close()` pair should be synchronous.

Reply:
Yeah, this was a typo, will address.

Reply:
Actually I think I was wrong in my assessment — buffered channel reads will block when the channel is empty, so what you had would indeed work, although the async wait/close is still technically unnecessary